File: apps/api/AssetManagement.Api/Controllers/CertificateTypesController.cs
using AssetManagement.Api.Data;
using AssetManagement.Api.DTOs;
using AssetManagement.Api.Models;
using AssetManagement.Api.Models.Enums;
using AssetManagement.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using EntityTypeEnum = AssetManagement.Api.Models.Enums.EntityType;

namespace AssetManagement.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/v1/[controller]")]
public class CertificateTypesController(AppDbContext db, IAuditService audit, ICurrentUserService currentUser) : ControllerBase
{
    [HttpGet]
    public async Task<ActionResult<PagedResponse<CertificateTypeDto>>> GetAll(
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 25,
        [FromQuery] string? search = null,
        [FromQuery] string sortBy = "name",
        [FromQuery] string sortDir = "asc")
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 100);

        var query = db.CertificateTypes
            .Where(t => !t.IsArchived)
            .Include(t => t.CustomFieldDefinitions)
            .AsQueryable();

        if (!string.IsNullOrWhiteSpace(search))
        {
            query = query.Where(t => EF.Functions.ILike(t.Name, $"%{search}%"));
        }

        var totalCount = await query.CountAsync();

        var desc = string.Equals(sortDir, "desc", StringComparison.OrdinalIgnoreCase);
        query = sortBy.ToLowerInvariant() switch
        {
            "description" => desc ? query.OrderByDescending(t => t.Description) : query.OrderBy(t => t.Description),
            "createdat" => desc ? query.OrderByDescending(t => t.CreatedAt) : query.OrderBy(t => t.CreatedAt),
            _ => desc ? query.OrderByDescending(t => t.Name) : query.OrderBy(t => t.Name),
        };

        var types = await query
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        var result = new PagedResponse<CertificateTypeDto>(
            types.Select(t => ToDto(t)).ToList(),
            page,
            pageSize,
            totalCount);

        return Ok(result);
    }

    [HttpGet("{id:guid}")]
    public async Task<ActionResult<CertificateTypeDto>> GetById(Guid id)
    {
        var type = await db.CertificateTypes
            .Include(t => t.CustomFieldDefinitions)
            .FirstOrDefaultAsync(t => t.Id == id);

        if (type is null) return NotFound();

        return Ok(ToDto(type));
    }

    [HttpGet("{id:guid}/customfields")]
    public async Task<ActionResult<List<CustomFieldDefinitionDto>>> GetCustomFields(Guid id)
    {
        var typeExists = await db.CertificateTypes.AnyAsync(t => t.Id == id);
        if (!typeExists) return NotFound();

        var definitions = await db.CustomFieldDefinitions
            .Where(d => d.CertificateTypeId == id && !d.IsArchived)
            .OrderBy(d => d.SortOrder)
            .Select(d => new CustomFieldDefinitionDto(
                d.Id, d.Name, d.FieldType.ToString(), d.Options, d.IsRequired, d.SortOrder))
            .ToListAsync();

        return Ok(definitions);
    }

    [HttpPost]
    public async Task<ActionResult<CertificateTypeDto>> Create(CreateCertificateTypeRequest request)
    {
        var type = new CertificateType
        {
            Id = Guid.NewGuid(),
            Name = request.Name,
            Description = request.Description
        };

        db.CertificateTypes.Add(type);

        if (request.CustomFields is { Count: > 0 })
        {
            foreach (var field in request.CustomFields)
            {
                if (!Enum.TryParse<CustomFieldType>(field.FieldType, out var fieldType))
                    return BadRequest(new { error = $"Invalid field type: {field.FieldType}" });

                var definition = new CustomFieldDefinition
                {
                    Id = Guid.NewGuid(),
                    EntityType = EntityTypeEnum.Certificate,
                    CertificateTypeId = type.Id,
                    Name = field.Name,
                    FieldType = fieldType,
                    Options = field.Options,
                    IsRequired = field.IsRequired,
                    SortOrder = field.SortOrder
                };
                db.CustomFieldDefinitions.Add(definition);
            }
        }

        await db.SaveChangesAsync();

        await audit.LogAsync(new AuditEntry(
            Action: "Created",
            EntityType: "CertificateType",
            EntityId: type.Id.ToString(),
            EntityName: type.Name,
            Details: $"Created certificate type \"{type.Name}\"" +
                (request.CustomFields is { Count: > 0 } ? $" with {request.CustomFields.Count} custom field(s)" : ""),
            ActorId: currentUser.UserId,
            ActorName: currentUser.UserName));

        await db.Entry(type).Collection(t => t.CustomFieldDefinitions).LoadAsync();

        return CreatedAtAction(nameof(GetById), new { id = type.Id }, ToDto(type));
    }

    [HttpPut("{id:guid}")]
    public async Task<ActionResult<CertificateTypeDto>> Update(Guid id, UpdateCertificateTypeRequest request)
    {
        var type = await db.CertificateTypes
            .Include(t => t.CustomFieldDefinitions)
            .FirstOrDefaultAsync(t => t.Id == id);

        if (type is null) return NotFound();

        type.Name = request.Name;
        type.Description = request.Description;
        type.UpdatedAt = DateTime.UtcNow;

        if (request.CustomFields is not null)
        {
            var existing = type.CustomFieldDefinitions.Where(d => !d.IsArchived).ToList();
            var requestIds = request.CustomFields
                .Where(f => f.Id.HasValue)
                .Select(f => f.Id!.Value)
                .ToHashSet();

            foreach (var def in existing)
            {
                if (!requestIds.Contains(def.Id))
                {
                    def.IsArchived = true;
                }
            }

            foreach (var field in request.CustomFields)
            {
                if (!Enum.TryParse<CustomFieldType>(field.FieldType, out var fieldType))
                    return BadRequest(new { error = $"Invalid field type: {field.FieldType}" });

                if (field.Id.HasValue)
                {
                    var def = existing.FirstOrDefault(d => d.Id == field.Id.Value);
                    if (def is not null)
                    {
                        def.Name = field.Name;
                        def.FieldType = fieldType;
                        def.Options = field.Options;
                        def.IsRequired = field.IsRequired;
                        def.SortOrder = field.SortOrder;
                    }
                }
                else
                {
                    var definition = new CustomFieldDefinition
                    {
                        Id = Guid.NewGuid(),
                        EntityType = EntityTypeEnum.Certificate,
                        CertificateTypeId = type.Id,
                        Name = field.Name,
                        FieldType = fieldType,
                        Options = field.Options,
                        IsRequired = field.IsRequired,
                        SortOrder = field.SortOrder
                    };
                    db.CustomFieldDefinitions.Add(definition);
                }
            }
        }

        await db.SaveChangesAsync();

        await audit.LogAsync(new AuditEntry(
            Action: "Updated",
            EntityType: "CertificateType",
            EntityId: type.Id.ToString(),
            EntityName: type.Name,
            Details: $"Updated certificate type \"{type.Name}\"",
            ActorId: currentUser.UserId,
            ActorName: currentUser.UserName));

        await db.Entry(type).Collection(t => t.CustomFieldDefinitions).LoadAsync();

        return Ok(ToDto(type));
    }

    [HttpDelete("{id:guid}")]
    public async Task<IActionResult> Archive(Guid id)
    {
        var type = await db.CertificateTypes.FindAsync(id);
        if (type is null) return NotFound();

        type.IsArchived = true;
        type.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();

        await audit.LogAsync(new AuditEntry(
            Action: "Archived",
            EntityType: "CertificateType",
            EntityId: type.Id.ToString(),
            EntityName: type.Name,
            Details: $"Archived certificate type \"{type.Name}\"",
            ActorId: currentUser.UserId,
            ActorName: currentUser.UserName));

        return NoContent();
    }

    private static CertificateTypeDto ToDto(CertificateType t) => new(
        t.Id, t.Name, t.Description,
        t.IsArchived, t.CreatedAt, t.UpdatedAt,
        t.CustomFieldDefinitions
            .Where(d => !d.IsArchived)
            .OrderBy(d => d.SortOrder)
            .Select(d => new CustomFieldDefinitionDto(
                d.Id, d.Name, d.FieldType.ToString(), d.Options, d.IsRequired, d.SortOrder))
            .ToList());
}

File: apps/api-kt/src/main/kotlin/com/assetmanagement/api/controller/CertificateTypesController.kt
package com.assetmanagement.api.controller

import com.assetmanagement.api.dto.*
import com.assetmanagement.api.model.CertificateType
import com.assetmanagement.api.model.CustomFieldDefinition
import com.assetmanagement.api.model.enums.CustomFieldType
import com.assetmanagement.api.model.enums.EntityType
import com.assetmanagement.api.repository.CertificateTypeRepository
import com.assetmanagement.api.repository.CustomFieldDefinitionRepository
import com.assetmanagement.api.service.AuditEntry
import com.assetmanagement.api.service.AuditService
import com.assetmanagement.api.service.CurrentUserService
import jakarta.persistence.criteria.Predicate
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.domain.Specification
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI
import java.time.Instant
import java.util.*

@RestController
@RequestMapping("/api/v1/certificatetypes")
class CertificateTypesController(
    private val certificateTypeRepository: CertificateTypeRepository,
    private val customFieldDefinitionRepository: CustomFieldDefinitionRepository,
    private val auditService: AuditService,
    private val currentUserService: CurrentUserService
) {

    @GetMapping
    fun getAll(
        @RequestParam(defaultValue = "1") page: Int, @RequestParam(defaultValue = "25") pageSize: Int,
        @RequestParam(required = false) search: String?, @RequestParam(defaultValue = "name") sortBy: String,
        @RequestParam(defaultValue = "asc") sortDir: String
    ): ResponseEntity<PagedResponse<CertificateTypeDto>> {
        val p = maxOf(1, page); val ps = pageSize.coerceIn(1, 100)
        val spec = Specification<CertificateType> { root, _, cb ->
            val preds = mutableListOf<Predicate>()
            preds.add(cb.equal(root.get<Boolean>("isArchived"), false))
            if (!search.isNullOrBlank()) preds.add(cb.like(cb.lower(root.get("name")), "%${search.lowercase()}%"))
            cb.and(*preds.toTypedArray())
        }
        val dir = if (sortDir.equals("desc", ignoreCase = true)) Sort.Direction.DESC else Sort.Direction.ASC
        val sort = Sort.by(dir, when (sortBy.lowercase()) { "createdat" -> "createdAt"; "description" -> "description"; else -> "name" })
        val result = certificateTypeRepository.findAll(spec, PageRequest.of(p - 1, ps, sort))
        return ResponseEntity.ok(PagedResponse(result.content.map { it.toDto() }, p, ps, result.totalElements))
    }

    @GetMapping("/{id}")
    fun getById(@PathVariable id: UUID): ResponseEntity<CertificateTypeDto> {
        val type = certificateTypeRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(type.toDto())
    }

    @GetMapping("/{id}/customfields")
    fun getCustomFields(@PathVariable id: UUID): ResponseEntity<Any> {
        if (!certificateTypeRepository.existsById(id)) return ResponseEntity.notFound().build()
        val defs = customFieldDefinitionRepository.findByCertificateTypeIdAndIsArchivedFalse(id)
            .sortedBy { it.sortOrder }.map { CustomFieldDefinitionDto(it.id, it.name, it.fieldType.name, it.options, it.isRequired, it.sortOrder) }
        return ResponseEntity.ok(defs)
    }

    @PostMapping
    fun create(@RequestBody request: CreateCertificateTypeRequest): ResponseEntity<Any> {
        val type = CertificateType(name = request.name, description = request.description)
        certificateTypeRepository.save(type)
        request.customFields?.forEach { field ->
            val ft = runCatching { CustomFieldType.valueOf(field.fieldType) }.getOrNull()
                ?: return ResponseEntity.badRequest().body(mapOf("error" to "Invalid field type: ${field.fieldType}"))
            customFieldDefinitionRepository.save(CustomFieldDefinition(entityType = EntityType.Certificate, certificateTypeId = type.id,
                name = field.name, fieldType = ft, options = field.options, isRequired = field.isRequired, sortOrder = field.sortOrder))
        }
        auditService.log(AuditEntry("Created", "CertificateType", type.id.toString(), type.name,
            "Created certificate type \"${type.name}\"", currentUserService.userId, currentUserService.userName))
        return ResponseEntity.created(URI("/api/v1/certificatetypes/${type.id}")).body(certificateTypeRepository.findById(type.id).get().toDto())
    }

    @PutMapping("/{id}")
    fun update(@PathVariable id: UUID, @RequestBody request: UpdateCertificateTypeRequest): ResponseEntity<Any> {
        val type = certificateTypeRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        type.name = request.name; type.description = request.description; type.updatedAt = Instant.now()
        if (request.customFields != null) {
            val existing = type.customFieldDefinitions.filter { !it.isArchived }
            val requestIds = request.customFields.mapNotNull { it.id }.toSet()
            existing.forEach { if (it.id !in requestIds) it.isArchived = true }
            request.customFields.forEach { field ->
                val ft = runCatching { CustomFieldType.valueOf(field.fieldType) }.getOrNull()
                    ?: return ResponseEntity.badRequest().body(mapOf("error" to "Invalid field type: ${field.fieldType}"))
                if (field.id != null) {
                    existing.find { it.id == field.id }?.apply { name = field.name; fieldType = ft; options = field.options; isRequired = field.isRequired; sortOrder = field.sortOrder }
                } else {
                    type.customFieldDefinitions.add(CustomFieldDefinition(entityType = EntityType.Certificate, certificateTypeId = type.id,
                        name = field.name, fieldType = ft, options = field.options, isRequired = field.isRequired, sortOrder = field.sortOrder))
                }
            }
        }
        certificateTypeRepository.save(type)
        auditService.log(AuditEntry("Updated", "CertificateType", type.id.toString(), type.name,
            "Updated certificate type \"${type.name}\"", currentUserService.userId, currentUserService.userName))
        return ResponseEntity.ok(certificateTypeRepository.findById(type.id).get().toDto())
    }

    @PostMapping("/bulk-archive")
    fun bulkArchive(@RequestBody request: BulkArchiveRequest): ResponseEntity<BulkActionResponse> {
        var s = 0; var f = 0
        request.ids.forEach { id -> val t = certificateTypeRepository.findById(id).orElse(null); if (t == null || t.isArchived) { f++; return@forEach }
            t.isArchived = true; t.updatedAt = Instant.now(); certificateTypeRepository.save(t)
            auditService.log(AuditEntry("Archived", "CertificateType", t.id.toString(), t.name, "Bulk archived", currentUserService.userId, currentUserService.userName)); s++ }
        return ResponseEntity.ok(BulkActionResponse(s, f))
    }

    @DeleteMapping("/{id}")
    fun archive(@PathVariable id: UUID): ResponseEntity<Any> {
        val t = certificateTypeRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        t.isArchived = true; t.updatedAt = Instant.now(); certificateTypeRepository.save(t)
        auditService.log(AuditEntry("Archived", "CertificateType", t.id.toString(), t.name,
            "Archived certificate type \"${t.name}\"", currentUserService.userId, currentUserService.userName))
        return ResponseEntity.noContent().build()
    }

    private fun CertificateType.toDto() = CertificateTypeDto(id, name, description, isArchived, createdAt, updatedAt,
        customFieldDefinitions.filter { !it.isArchived }.sortedBy { it.sortOrder }
            .map { CustomFieldDefinitionDto(it.id, it.name, it.fieldType.name, it.options, it.isRequired, it.sortOrder) })
}
